server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByFillWithValueFilterDataSet.java

```diff
@@ -0,0 +1,438 @@
+      }
+    } finally {
+      StorageEngine.getInstance().mergeUnLock(list);
+    }
+  }
+
+  private TimeGenerator getTimeGenerator(QueryContext context, RawDataQueryPlan queryPlan)
+      throws StorageEngineException {
+    return new ServerTimeGenerator(context, queryPlan);
+  }
+
+  private IReaderByTimestamp getReaderByTime(
+      PartialPath path, RawDataQueryPlan queryPlan, TSDataType dataType, QueryContext context)
+      throws StorageEngineException, QueryProcessException {
+    return new SeriesReaderByTimestamp(
+        path,
+        queryPlan.getAllMeasurementsInDevice(path.getDevice()),
+        dataType,
+        context,
+        QueryResourceManager.getInstance().getQueryDataSource(path, context, null),
+        null,
+        ascending);
+  }
+
+  private void replaceGroupByFilter(IExpression expression, Filter timeFilter)
+      throws QueryProcessException, IllegalPathException {
+    if (expression instanceof SingleSeriesExpression) {
+      ((SingleSeriesExpression) expression)
+          .setSeriesPath(
+              new PartialPath(((SingleSeriesExpression) expression).getSeriesPath().getFullPath()));
+      if (((SingleSeriesExpression) expression).getFilter() instanceof GroupByFilter) {
+        ((SingleSeriesExpression) expression).setFilter(timeFilter);
+      } else if (((SingleSeriesExpression) expression).getFilter() instanceof BinaryFilter) {
+        if (((BinaryFilter) ((SingleSeriesExpression) expression).getFilter()).getLeft()
+            instanceof GroupByFilter) {
+          ((BinaryFilter) ((SingleSeriesExpression) expression).getFilter()).setLeft(timeFilter);
+        } else if (((BinaryFilter) ((SingleSeriesExpression) expression).getFilter()).getRight()
+            instanceof GroupByFilter) {
+          ((BinaryFilter) ((SingleSeriesExpression) expression).getFilter()).setRight(timeFilter);
+        }
+      } else {
+        throw new QueryProcessException("unknown filter type, can't replace group by filter");
+      }
+    } else if (expression instanceof BinaryExpression) {
+      replaceGroupByFilter(((BinaryExpression) expression).getLeft(), timeFilter);
+      replaceGroupByFilter(((BinaryExpression) expression).getRight(), timeFilter);
+    } else {
+      throw new QueryProcessException("unknown expression type, can't replace group by filter");
+    }
+  }
+
+  // get new expression that can query extra range
+  private IExpression getNewExpression(GroupByTimeFillPlan groupByTimeFillPlan, Filter timeFilter)
+      throws QueryProcessException {
+    IExpression newExpression = groupByTimeFillPlan.getExpression().clone();
+    try {
+      replaceGroupByFilter(newExpression, timeFilter);
+    } catch (IllegalPathException ignore) {
+      // ignored
+    }
+    return newExpression;
+  }
+
+  /* Init extra path executors to query data outside the original group by query */
+  private void initExtraGenerators(QueryContext context, GroupByTimeFillPlan groupByTimeFillPlan)
```

Comment:
rename it to `initExtraReadersByTimestamp`

Reply:
done